internal/fnruntime/container.go

```diff
@@ -80,6 +81,10 @@ type ContainerFn struct {
 // It reads the input from the given reader and writes the output
 // to the provided writer.
 func (f *ContainerFn) Run(reader io.Reader, writer io.Writer) error {
+	// check for docker availability
+	if err := dockerCmdAvailable(); err != nil {
```

Comment:
I understand the motivation to move it to a common layer here but the approach has downsides:

- the check is performed on every function run now, so if pipeline has 10 functions, `docker version` will be invoked 10 times so it adds to the latency
- Since the check is done at the lowest layer instead of the first thing when `fn eval` or `fn render` is called, so it doesn't fail fast.

So having a common utility function to check availability of docker and invoking it in the `fn eval` and `fn render` command runners as pre-run step seems to avoid both the above problem with a little downside of repeating the call in both the places.

Reply:
got it, thanks. Moved it out.